From d6ea8814e7f189a4c8252630389bfd23d580d253 Mon Sep 17 00:00:00 2001
From: ThatApplePieGuy <necrozma999@gmail.com>
Date: Sun, 1 Feb 2026 18:16:25 -0700
Subject: [PATCH] Support io_uring


diff --git a/pom.xml b/pom.xml
index dcdc0fe8b..794170730 100644
--- a/pom.xml
+++ b/pom.xml
@@ -85,6 +85,25 @@
             <artifactId>netty-transport-native-epoll</artifactId>
             <classifier>linux-x86_64</classifier>
         </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-classes-io_uring</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-io_uring</artifactId>
+            <classifier>linux-aarch_64</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-io_uring</artifactId>
+            <classifier>linux-riscv64</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-io_uring</artifactId>
+            <classifier>linux-x86_64</classifier>
+        </dependency>
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-transport-native-unix-common</artifactId>
diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
index 8760ecb66..832de2e3f 100644
--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
@@ -222,4 +222,9 @@ public class PaperConfig
     private static void trackPluginScoreboards() {
         trackPluginScoreboards = getBoolean("settings.track-plugin-scoreboards", false);
     }
+
+    public static boolean useIoUring = true;
+    private static void useIoUring() {
+        useIoUring = getBoolean("settings.use-io-uring", useIoUring);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/ServerConnection.java b/src/main/java/net/minecraft/server/ServerConnection.java
index acfa52d6e..f83547550 100644
--- a/src/main/java/net/minecraft/server/ServerConnection.java
+++ b/src/main/java/net/minecraft/server/ServerConnection.java
@@ -49,9 +49,9 @@ public class ServerConnection {
             return this.a();
         }
     };
-    public static final LazyInitVar<LocalEventLoopGroup> c = new LazyInitVar() {
-        protected LocalEventLoopGroup a() {
-            return new LocalEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Local Server IO #%d").setDaemon(true).build());
+    public static final LazyInitVar<io.netty.channel.MultiThreadIoEventLoopGroup> SERVER_IO_URING_EVENT_GROUP = new LazyInitVar() {
+        protected io.netty.channel.MultiThreadIoEventLoopGroup a() {
+            return new io.netty.channel.MultiThreadIoEventLoopGroup(0, new ThreadFactoryBuilder().setNameFormat("Netty io_uring Server IO #%d").setDaemon(true).build(), io.netty.channel.uring.IoUringIoHandler.newFactory());
         }
 
         protected Object init() {
@@ -92,7 +92,14 @@ public class ServerConnection {
             Class oclass;
             LazyInitVar lazyinitvar;
 
-            if (Epoll.isAvailable() && this.f.ai()) {
+            if (com.destroystokyo.paper.PaperConfig.useIoUring && io.netty.channel.uring.IoUring.isAvailable()) {
+                // Paper start - Unix domain socket support
+                if (address instanceof io.netty.channel.unix.DomainSocketAddress) oclass = io.netty.channel.uring.IoUringServerDomainSocketChannel.class;
+                else oclass = io.netty.channel.uring.IoUringServerSocketChannel.class;
+                // Paper end
+                lazyinitvar = ServerConnection.SERVER_IO_URING_EVENT_GROUP;
+                ServerConnection.e.info("Using io_uring channel type");
+            } else if (Epoll.isAvailable() && this.f.ai()) {
                 // Paper start - Unix domain socket support
                 if (address instanceof io.netty.channel.unix.DomainSocketAddress) {
                     oclass = io.netty.channel.epoll.EpollServerDomainSocketChannel.class;
-- 
2.54.0

